### PR TITLE
Remove max memory as a user setting available in GUI, move to game_engine.properties

### PR DESCRIPTION
**Pull request by DanVanAtta**
Notable updates:
- remove GUI support for updating java xmx (max memory)
- remove the 'xmx' property in `system.ini` and add an equivalent `max_memory` property to `game_engine.properties`
- simplify the check for if memory is set and move the responsibility for relaunching the game to apply custom memory settings. Have GameRunner do the launch of a new game and simply return from main(), avoiding a call to System.exit. Second, avoid the 'within 20% of current memory', if there is an override set, then use it.
- remove the 'xmx use default' property in system.ini, instead we'll rely on `max_memory` being set in properties or not set. This removes the last property in `system.ini`, the file was removed along with the related build config.

### CODE REVIEW

**Comment by Cernelius** [comment]
So, I don't understand most of this, and I've already expressed my strong positive opinion in the existence of the Max Memory Setting in the past.
I'm fine with any as long I can still set it. Tho I surely believe it would be better to keep a way for anyone to set it easily, like the current "Set Max Memory Usage".
**I don't really understand, but will it be still possible to like set it at 4096 MB, and how?**
**Can I go in the `game_engine.properties`, and set it at whatever value just like I could have done in `system.ini`, right?**
In general, as I already said in the past, I think everyone should be able (as they currently are) to not be limited by a set max memory, but go over it, if wished (it is relevant for maps with a big / slow xml and for playing with Hard AI).
The main item is that you can't say how much xml heavy maps will be made in the future (I've under work some with a big xml), so a setting for having the max memory as high as one wishes is needed, not to forcefully limit what one can possibly do (if the xml is very heavy you can even hit the max memory and be unable to start the map at all).
So, what I'm saying is that I don't really understand what this is about, but as long as the functionality to increase the max memory as wished is kept, then it is fine (otherwise, I strongly suggest not to remove this feature). Also, I suggest it being easy to manage (like now, doable in Engine Preferences/Set Max Memory Usage), but this is a plus; main item is being always at least able to do it at all (always able to not have to have a default limit imposed).

**Comment by Cernelius** [comment]
I mostly don't understand much, so if the matter is just doing about the same thing but with this game_engine.properties instead of system.ini; then of course nevermind me (I've no clue about why one or the other, and don't care, as long as you don't have an unavoidable limit on the max memory that you can hit).

**Comment by Cernelius** [comment]
Ok, I've missed the link and just read at:
https://github.com/triplea-game/triplea/issues/836
From what I understand the changes are:
1. The ability to set it at whatever levels in any cases is fully kept, just moved from editing `system.ini` to `game_engine.properties`, with a text editor.
2. The ability to set it via the Engine is removed totally; users will have to go in their `game_engine.properties` and change the value, if they want to change their max memory.

Ok, if this is correct, then I can't have a strong argument against this change, tho I don't like the removal of it as a setting that you can manage in the engine.
Number 1 is the most important, and if it is kept, then it is fine enough.
I really believe that number 2 is important too, but mostly in perspective (you cannot know how many heavy maps will be made and how much popular will they be), since currently there are not many or popular heavy xml maps and increasing the memory for Hard AI play might be of limited impact, so probably not many people use that feature, that I surely suggest to keep anyway. I guess it will have to be rediscussed if, in the future, there will be more maps for which users may want to increase their memory (and not all might be power user enough to do it easily without something integrated in the engine for doing it).

So, to sum it up, as long as you can change the max setting in any ways that is the main thing. I'm surely sad of the removal of the option in the engine, and surely suggest not to, but, sadly, I'm not able to present a strong argument in its favour, based on maps currently available in the repository, and likely very few people use it, right now.
I recall there were cases in the past of increasing it for playing maps like TWW or Domination with HardAI, where users reported better performance, but I personally rarely play AI, so I can't tell myself. Maybe @ron-murhammer recalls those cases. My main item is not limiting how much heavy a map can possibly be, mainly referring to the xml (quantity of triggers etc.).

**Comment by Cernelius** [comment]
So to really sum up what I was saying is that if I understand correctly that you can still set it, then I can't be strongly against, or at least not to make a strong actual argument, related to my use of maps currently in the repository, but I would still say, please, keep it as something that anyone can easily set, if they happen to require it to play a game that needs it (that makes you hit the max memory and gives problem, or even crash, because of too heavy for it; also the need increases while adding up played rounds).

**Comment by RoiEXLab** [comment]
@DanVanAtta I'm not sure if moving this option to game:engine.properties is the best Idea.
The memory setting is a java argument, so it needs to be applied before TripleA is being launched.
Without the install4j launcher this is really easy, one would just add `-Xmx512m` as command line arg before launching triplea or create a batch/bash script which does exactly that: `java -Xmx512M -Xms512M -jar /path/to/triplea.jar`
I just noticed that install4j makes this really easy as well. In every TripleA directory (On windows `C:\Program Files\TripleA_<version>\`) there is an TripleA.vmoptions file which accepts command line args for the java virtual machine e.g. `-Xmx512M`.
On would just need to add this line to this file and everything would work as expected.
I believe this is far better than parsing the game_engine-properties file and then relaunch triplea if any settings need to be made before invoking triplea.
Ideally the `GameRunner#startNewTripleA` and `TripleA#startGame` methods should be removed entirely as creating multiple processes makes debugging incredibly difficult. (When launching triplea with any IDE the stacktrace will no longer be shown and the application will be marked as terminated, because it actually was), but I know they are currently used to host games etc., so removing them would be the easiest task.

**Comment by DanVanAtta** [comment]
> I'm not sure if moving this option to game_engine.properties is the best Idea.
The memory setting is a java argument, so it needs to be applied before TripleA is being launched.

Whether the value is in a file called `system.ini` or a file called `game_engine.properties`, we still have the existing behavior where the game checks if max memory was set and then relaunches itself. I would rather this not be the existing behavior, but it is for now.. In this update what we are doing is removing the GUI option to set max memory, and moving the value from `system.ini` to `game_engine.properties`. This way we have one fewer config files and avoid having to fix the GUI. The GUI for this property right now is a very long scary warning message, that may as well be buried in a config. It is set up so only power users would modify it, it seems okay to have power users modify a properties file IMO, so we would just as well hide the UI option from all the other users. It's good for us to remove/hide controls which we do not expect normal users to updates, but we do still allow power users to update the memory option. 

My real problem with the max memory setting is that it is a value that should be set optimally, and not only be available to power users. Regardless of that, the above is some context and motivation of what we are doing on this branch and a bit more behind why.

**Comment by DanVanAtta** [comment]
@RoiEXLab I agree with you, but please note my preceding comment, the goal here is to kill a GUI option and remove the `system.ini` config file so we only have one. I think the direction you mention is what we should be doing - the installer should be setting this up for us. The only issue then is calcuating a good value, regardless that is beyond this PR.

**Comment by DanVanAtta** [comment]
@Cernelius preceding comment should also address your concern, the ability to set max memory is not going away entirely.

**Comment by RoiEXLab** [comment]
@DanVanAtta I will open a PR on this issue, once this PR is getting merged...
This may be a bad example, but when it comes to Minecraft, almost everyone wanted to assign more ram, because it "could make your game faster". At least that is what many think... The minority are probably power-users, but in most cases they are able to find a tutorial obn youtube showing how simple it is to add those command line params.
We should probably add a page on the website explaining this procedure, once I created this PR

File: src/main/java/games/strategy/engine/framework/system/Memory.java [hunk 1]
```diff
@@ -1,101 +1,50 @@
...
-    final long currentMaxMemory = Runtime.getRuntime().maxMemory();
-    System.out.println("Current max memory: " + currentMaxMemory + ";  and new xmx should be: " + xmx);
-    final long diff = Math.abs(currentMaxMemory - xmx);
-    // Runtime.maxMemory is never accurate, and is usually off by 5% to 15%,
-    // so if our difference is less than 22% we should just ignore the difference
-    if (diff <= xmx * 0.22) {
-      return;
+
+    final String maxMemorySetting = ClientContext.propertyReader().readProperty(GameEngineProperty.MAX_MEMORY);
+    if (Strings.nullToEmpty(maxMemorySetting).isEmpty() || Long.parseLong(maxMemorySetting) < 32L) {
```
Comment (ssoloff) [inline src/main/java/games/strategy/engine/framework/system/Memory.java]: Just to make sure I'm reading this correctly...  The minimum user max memory setting is 32 MiB?  Otherwise, we call them crazy and ignore what they provided?

**Comment by Cernelius** [comment]
In case it got overlooked, what I was saying is that, in my opinion, the main item behind being able not to be limited by an imposed max memory is not to make the game faster in maps that would run anyways but the fact that you can possibly have maps with an XML so heavy that you would not be able to play them without increasing the ram assigned (they would hit the max and not start or stop).
Thus, as it is, a not settable max memory would impose max limits to mapmaking, in terms of how many triggers etc. you can have (think about a bigger version of war of the relics).
Anyways, unless the max go away entirely (my preference, but I don't know if it is feasible), I believe there should always be a way for all users not to have an imposed max memory TripleA can't forcibly go over.

This is only a sum-up of what I already said, also outside of this pull request; not really reopening the discussion.
I confirm I understand that, with these changes, you can still reset the max however you wish to, by changing the settings with a text editor (IMO good), but not with an option in the engine anymore (IMO bad), and I've nothing else to say than what I already did.

So, to sum up, I'm sad that the option in the engine is going to be removed, but I'm glad that at least the txt editable setting is kept.

**Comment by Cernelius** [comment]
Sorry for repeating, but, just to be clear, as I said several times here and in several issues about it, I have not changed my opinion (expressed above) and I still surely suggest against removing the possibility of setting the max memory as an engine option (of course, still allowing setting it via something in the installation I believe is much better than removing it entirely, no doubt).

The reasons are those I've already said, here an in other issues. Just confirming I still believe in what I said.

This is just to record that I confirm I'm personally against removing partially or totally this option (of course, I believe removing it from the interface only is much better than removing it totally).

I realise I can't stop you from doing it nor I can cast a vote on it.

**Comment by RoiEXLab** [comment]
@Cernelius I played around with a couple install4j settings, this is what we can do when moving this setting to the generated vmoptions file:
We can specify a default RAM size when installing.
This means, if someone created a map which required more than the standard 1G we could increase this limit to make this map playable.
'Power Users' could however still fiddle around with those settings on their own.

File: game_engine.properties [hunk 2]
```diff
@@ -10,3 +10,6 @@ Map_List_File = http://raw.githubusercontent.com/triplea-game/triplea/master/tri
 lobby_properties_file_url = http://raw.githubusercontent.com/triplea-game/triplea/master/lobby_server.yaml?raw=true
 lobby_properties_file_backup = lobby_server.yaml
 
+# Override setting for max memory in MB, leave blank to use default.
+# It is recommended to use the default, setting a value too high or too low can cause TripleA to be unstable.
+max_memory=1024
```
Comment (DanVanAtta) [inline game_engine.properties]: Good catch, that was left over from testing. Updated in: 676a2f3, property is now blank to match the commentary.

File: src/main/java/games/strategy/engine/framework/system/Memory.java (hunk 1 above)
Comment (DanVanAtta) [inline src/main/java/games/strategy/engine/framework/system/Memory.java]: Yup : )
I could not really decide on what a reasonable minimum is. Non-zero is bare-minimum, but '1MB' does not make sense either. The game AFAIK really wants 128MB~256MB and upwards of half a gig. 16MB culd make sense, but it is still crazy.. 

Given the above hemming and hawwing.. Let's go with `< 1L`
- a user override is a user override. The user will know hopefully that `1L` is just as crazy as `32MB`, but less than `1` is just invalid..

**Comment by DanVanAtta** [comment]
I think that addresses everything specific. @ssoloff @RoiEXLab , anything not addressed?

@RoiEXLab I agree that TripleA should not handle Xmx the way it does. That is not the goal here though. Here we remove a GUI control that is for power users only. Since we would redo how the setting is done, the logic behind the GUI is not going to be used, the power-user only text is not re-usable, and the logic behind the feature would be updated. Having a power user only setting buried in the settings is the same as the same setting buried in a config. Meanwhile we simplify quite a bit in terms of code, and also we go from 2 client configuration files to 1. This should also make it easier to redo how we handle XMX since there are fewer dependencies on it, notably fewer system prop dependencies.

**Comment by ssoloff** [comment]
Latest changes look good.
